refactor(presence): Use branded JsonDeserialized type internally

tylerbutler opened the pull request:
Updates the presence internals to use an opaque/branded internal type when passing around JSON objects. There are two functions to make JSON types opaque/not opaque, as well as a helper that casts from an opaque JSON type to a `DeeplyReadonly<JsonDeserialized<T>>` object since that is what we often need.

Externally, the API remains the same; the internal types and functions are marked `@system` and objects are converted between the types at the API boundaries.

There were API level mismatches with the Notifications changes, so I marked `InternalUtilityTypes` as alpha. Is there a better change? Similarly, the core-interfaces types needed to be beta since the presence APIs are marked as such.

tylerbutler committed:
typebrand the JsonDeserialized type

tylerbutler committed:
updates

tylerbutler committed:
compiles

jason-ha commented on packages/framework/presence/src/exposedUtilityTypes.ts:
```diff
@@ -76,4 +76,52 @@ export namespace InternalUtilityTypes {
+	 * @system
+	 */
+	// export type JsonDeserializedHandle<T> = Tagged<JsonDeserialized<T>, "JsonDeserialized">;
+	export type JsonDeserializedHandle<T> = T & JsonDeserializedBrand<T>;
```
Probably want to remove the `T &` because (1) it shouldn't be used as if it were actually `T` (it might not be) and (2) we don't expect to have any case where it is used as `T` (so don't give ourselves access to confirm).
It does mean we will need to cast through `unknown`, but that is fine.

jason-ha commented on packages/framework/presence/src/exposedUtilityTypes.ts:
```diff
@@ -76,4 +76,52 @@ export namespace InternalUtilityTypes {
+	 * @system
+	 */
+	export declare class JsonDeserializedBrand<T> {
+		private readonly JsonDeserialized: JsonDeserialized<T>;
```
We can probably drop the `JsonDeserialized<T>`. Just `T` is probably enough. Probably want to use the dummy accessor and iterator found in that reference PR's brand classes.

jason-ha commented on packages/framework/presence/src/latestMapValueManager.ts:
```diff
@@ -266,22 +271,24 @@ class ValueMapImpl<T, K extends string | number> implements StateMap<K, T> {
 		for (const [key, item] of objectEntries(this.value.items)) {
 			if (item.value !== undefined) {
-				callbackfn(asDeeplyReadonly(item.value), key, this);
+				const unbrandedValue = fromJsonDeserializedHandle(item.value);
+				callbackfn(asDeeplyReadonly(unbrandedValue), key, this);
```
Can probably combine both of these pass-through helper functions.

jason-ha commented on packages/framework/presence/src/latestMapValueManager.ts:
```diff
@@ -266,22 +271,24 @@ class ValueMapImpl<T, K extends string | number> implements StateMap<K, T> {
 	}
 	public get(key: K): DeepReadonly<JsonDeserialized<T>> | undefined {
-		return asDeeplyReadonly(this.value.items[key]?.value);
+		const data = this.value.items[key]?.value;
+		return data === undefined ? undefined : asDeeplyReadonly(fromJsonDeserializedHandle(data));
```
Kind of silly since those are both pass through anyway - should be able to adjust the helper to handle it.

tylerbutler committed:
Remove T &

tylerbutler committed:
feedback

tylerbutler committed:
rename types to behopefully clearer

tylerbutler committed:
feedback

tylerbutler committed:
cleanup

jason-ha committed:
feat(client-core-interfaces): opaque JSON placeholder types
Introduce `OpaqueJsonDeserialized`, and `OpaqueJsonSerializable` as "handles" for generic values.

Opaque Json types can be used when there is no direct need to work with a complex or unknown type, but still need to respect serializability. Common use is when a value that has passed through `JsonSerializable` or `JsonDeserialized` filter utilities needs to be stored, especially in a generic context. The Opaque types are recognized by `JsonSerializable` and `JsonDeserialized` allowing those stored captures to themselves be part of flow that checks for serializability or deserialization.

Test cases:
  JsonDeserialized
    positive compilation tests
      fully supported object types are preserved
        ✔ OpaqueJsonDeserialized<{number:number}>
      partially supported object types are modified
        OpaqueJsonSerialized becomes OpaqueJsonDeserialized counterpart
          ✔ OpaqueJsonSerializable<{number:number}> becomes OpaqueJsonDeserialized<{number:number}>
          ✔ OpaqueJsonSerializable<{number:number}>&OpaqueJsonDeserialized<{number:number}> becomes OpaqueJsonDeserialized<{number:number}>
        opaque Json types requiring extra allowed types have extras removed
          ✔ opaque serializable object with `bigint`
          ✔ opaque deserialized object with `bigint`
          ✔ opaque serializable and deserialized object with `bigint`
          ✔ opaque serializable object with number array expecting `bigint` support
          ✔ opaque deserialized object with number array expecting `bigint` support
          ✔ opaque serializable and deserialized object with number array expecting `bigint` support
    special cases
      using alternately allowed types
        are preserved
          ✔ opaque serializable object with `bigint`
          ✔ opaque deserialized object with `bigint`
          ✔ opaque serializable and deserialized object with `bigint`

  JsonSerializable
    positive compilation tests
      supported object types
        opaque Json types
          ✔ opaque serializable object
          ✔ opaque deserialized object
          ✔ opaque serializable and deserialized object
    negative compilation tests
      unsupported types cause compiler error
        opaque Json types requiring extra allowed types
          ✔ opaque serializable object with `bigint`
          ✔ opaque deserialized object with `bigint`
          ✔ opaque serializable and deserialized object with `bigint`
          ✔ opaque serializable object with number array expecting `bigint` support
          ✔ opaque deserialized object with number array expecting `bigint` support
          ✔ opaque serializable and deserialized object with number array expecting `bigint` support

  OpaqueJsonSerializable and OpaqueJsonDeserialized
    positive compilation tests
      ✔ OpaqueJsonSerializable is covariant (more specific is assignable to general)
      ✔ OpaqueJsonDeserialized is covariant (more specific is assignable to general)
      ✔ OpaqueJsonSerializable & OpaqueJsonDeserialized is covariant (more specific is assignable to general)
      in a generic context
        ✔ OpaqueJsonDeserialized assignability varies with JsonDeserialized Options
        ✔ OpaqueJsonSerializable assignability varies with JsonSerializable Options
        ✔ OpaqueJsonSerializable may be forwarded as JsonSerializable
        ✔ OpaqueJsonDeserialized may be returned to JsonDeserialized
        ✔ OpaqueJsonSerializable & OpaqueJsonDeserialized may be returned to JsonSerializable & JsonDeserialized
        ✔ OpaqueJsonSerializable & OpaqueJsonDeserialized may be forwarded as JsonSerializable
        ✔ OpaqueJsonSerializable & OpaqueJsonDeserialized may be returned as JsonDeserialized
    negative compilation tests
      ✔ OpaqueJsonSerializable is covariant (more general is NOT assignable to specific)
      ✔ OpaqueJsonDeserialized is covariant (more general is NOT assignable to specific)
      ✔ OpaqueJsonSerializable & OpaqueJsonDeserialized is covariant (more specific is assignable to general)

tylerbutler committed:
Merge branch 'main' into presence-type-branding

tylerbutler committed:
cleanup

tylerbutler committed:
fix

tylerbutler committed:
Merge branch 'main' into presence-type-branding

tylerbutler committed:
rm invalid docs changes

tylerbutler committed:
formatting

tylerbutler committed:
tsignore some errors

tylerbutler committed:
wip

tylerbutler committed:
wip

tylerbutler committed:
Merge branch 'main' into presence-type-branding

tylerbutler committed:
wip

tylerbutler committed:
formatting

tylerbutler committed:
wip

tylerbutler committed:
Merge branch 'core/opaque-json-types' into presence-type-branding

tylerbutler committed:
revert container-runtime-defs changes

tylerbutler committed:
8 errors in 3 files

tylerbutler committed:
rename functions

tylerbutler committed:
8 errors, 3 files (still)

tylerbutler committed:
rename and reorg

jason-ha committed:
improvement(client-presence): use opaque json internally
Simplify internal unknown (generic) customer data that must be Json serializable using Opaque Json types

At boundaries of API ins and outs, Json filtered types are cast to/from Opaque Json types. See new helpers:
  - `fullySerializableToOpaqueJson`
  - `asDeserializedJson`
  - `asDeeplyReadonlyDeserializedJson`

Notifications utilities are updated and removal of `string &` for `NotificationListeners` allowed clean up of `@ts-ignore-error`. `NotificationsManagerImpl`'s `emit` implementation now just types `name` to `string`.'

jason-ha committed:
improvement(client-core-interfaces): list OpaqueJsonSerializable as degenerate option
Listing `OpaqueJsonSerializable<unknown>` for degenerate option allows a filtered and captured type to accept other declared `OpaqueJsonSerializable` types.

tylerbutler committed:
Merge branch 'main' into presence-type-branding

tylerbutler committed:
workarounds

tylerbutler committed:
Merge commit 'dbf0696243' into presence-type-branding

tylerbutler committed:
source compiles but tests don't

tylerbutler committed:
source compiles

jason-ha committed:
improvement(client-presence): reduce data in requirement to JsonSerializable
as we are confident that is sufficient. See #24751.

Update opaque cast helper to only require `JsonSerializable`.

tylerbutler committed:
revert

tylerbutler committed:
Merge branch 'core/promote-OpaqueJsonSerializable-for-degenerate-cases' into presence-type-branding

jason-ha committed:
test(client-presence): updates for Opaque Json use

Note: `testUtils.ts`'s `prepareConnectedPresence` has error where an `InboundClientJoinMessage` is supposed to satisfy `OutboundClientJoinMessage`.

jason-ha committed:
docs(client-presence): update API reports
after ignoring remaining type issue in presenceDatastoreManager.ts
`PresenceDatastore` does not satisfy `DatastoreMessageContent`

jason-ha committed:
test(client-presence): fix for JsonSerializable change

tylerbutler committed:
Merge branch 'presence/use-opaque-json' into presence-type-branding

tylerbutler committed:
use ts-expect-error instead of as any

tylerbutler committed:
Merge branch 'core/promote-OpaqueJsonSerializable-for-degenerate-cases' into presence-type-branding

tylerbutler committed:
remove ts-expect-error that is no longer triggered

jason-ha committed:
merge: 'main' into presence/use-opaque-json

jason-ha committed:
test(client-presence): correction after merge

jason-ha committed:
fix(client-presence): workaround "system:presence" incompatilities
 - `PresenceDatastore` -> `DatastoreMessageContent`
 - `InboundClientJoinMessage` -> `OutboundClientJoinMessage`

tylerbutler committed:
Merge branch 'main' into presence-type-branding

tylerbutler committed:
Merge branch 'presence/use-opaque-json' into presence-type-branding

tylerbutler committed:
test-case ts-expect-error

tylerbutler committed:
core-interfaces beta/system

tylerbutler committed:
presence beta/system

jason-ha committed:
Revert "fix(client-presence): workaround "system:presence" incompatilities"

This reverts commit 2aab49ee7d872035802fff304fc14bf884aac5da.

tylerbutler committed:
rename and cleanup

tylerbutler committed:
use new to/from opaque function names

tylerbutler committed:
remove beta flags in types where possible

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -172,12 +172,24 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
 		if (updateProviders.length > 3) {
 			updateProviders.length = 3;
 		}
+		// TODO: #??? investigate and address `PresenceDatastore` incompatibility with `DatastoreMessageContent`
```
Open a tracking issue.

tylerbutler commented on packages/framework/presence/src/test/testUtils.ts:
```diff
@@ -146,12 +147,35 @@ export function prepareConnectedPresence(
 
-	const expectedClientJoin: OutboundClientJoinMessage &
-		Partial<Pick<InboundClientJoinMessage, "clientId">> = generateBasicClientJoin(clock.now, {
+	// TODO: #??? investigate and address `InboundClientJoinMessage` incompatibility with `OutboundClientJoinMessage`
```
File an issue.

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -327,17 +339,32 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
 	}
 
 	private broadcastAllKnownState(): void {
+		// TODO: #??? investigate and address `PresenceDatastore` incompatibility with `DatastoreMessageContent`
```
Same issue as above.

tylerbutler committed:
add back beta

tylerbutler committed:
imports

jason-ha commented on packages/common/core-interfaces/src/internal.ts:
Looks like all of these are incorrect. Likely you imported from the wrong path somewhere. See the note on lines 15-18.

jason-ha committed:
fix(client-presence): remove opacity from system datastore

jason-ha committed:
fix(client-core-interfaces): support Opaque unknown thru JsonSerializable
Filtering the inner type when applying additional `JsonSerializable` filters broken the result down into `JsonTypeWith | Opaque` degenerate substitutes when `unknown` was sufficient already and would break the filter type checking.

Now preserve the Opaque type's `T` as-is and rely on changes in `Controls` to dictate incompatibility.

jason-ha committed:
fix(client-core-interfaces): support Opaque unknown thru JsonSerializable
Filtering the inner type when applying additional `JsonSerializable` filters broken the result down into `JsonTypeWith | Opaque` degenerate substitutes when `unknown` was sufficient already and would break the filter type checking.

Now preserve the Opaque type's `T` as-is and rely on changes in `Controls` to dictate incompatibility.

jason-ha committed:
improvement(client-presence): separate system:presence index from others
This is required to allow `system:presence` workspace's datastore to deviated from the common general workspace datastore.

This also removes test workarounds when specifying `clientToSessionId` records.

jason-ha committed:
fix(client-presence): separate system:presence index from others
This is required to allow `system:presence` workspace's datastore to deviated from the common general workspace datastore.

This also removes test workarounds when specifying `clientToSessionId` records.

tylerbutler committed:
Merge branch 'core/fix-Opaque-unknown-serializability' into presence-type-branding

tylerbutler committed:
Merge branch 'presence/improvement-separate-system_presence-from-general-datastore-typing' into presence-type-branding

tylerbutler committed:
wip

tylerbutler committed:
Merge branch 'main' into presence-type-branding

tylerbutler committed:
fix

tylerbutler committed:
Merge branch 'presence/use-opaque-json' into presence-type-branding

jason-ha committed:
merge: 'main' into presence/use-opaque-json

tylerbutler committed:
wip

tylerbutler committed:
Merge branch 'presence/use-opaque-json' into presence-type-branding

tylerbutler committed:
cleanup

tylerbutler committed:
possible fix

tylerbutler committed:
updates

jason-ha committed:
docs(client-presence): comment additions

jason-ha committed:
refactor(client-presence): rename helpers
- `serializableToOpaqueJson` -> `toOpaqueJson`
- `asDeserializedJson` -> `revealOpaqueJson`

tylerbutler committed:
Merge branch 'presence/use-opaque-json' into presence-type-branding

jason-ha committed:
revert(client-presence): unneeded code changes

tylerbutler committed:
Merge branch 'presence/use-opaque-json' into presence-type-branding

tylerbutler commented:
This change is superseded by #24772.
